### DISCUSSION TITLE
Realm (or any) Database Best Practices

### DISCUSSION THREAD

**Discussion by hendych** [MobileNativeFoundation/discussions#177] [General]
Hi everyone!
There are crashes often caused by Realm migration failures. (maybe other databases too?)
I want to know how do you manage your realm database in a big app especially [a super app](https://en.wikipedia.org/wiki/Super-app) such as:
- Do you create centralized database that will be used for all products?
- How do you manage the database schema migration? How do you handle the migration failures (rollback, error handling, app recovery)?
- How do you manage the realm schema versioning?

**Comment by philipphofmann** [comment]
I can share my experiences as an Android from my previous company [mySugr](https://www.mysugr.com/en/).

At some point, we introduced Realm and also had SQLite. The data stored in the two databases wasn't really connected, so we didn't have to merge the data often. I think it is fine two have different databases in your app if you have good reasons for it. If a particular feature has completely different requirements, which outweigh the overhead of an extra database, I think it's OK to add a new one. 

We managed database migration by having some good test data and writing integration tests running on simulators. Furthermore, we had QA testing the version updates from one app version to another manually to ensure it didn't break. 

We didn't use Realm on the backend, so we had no issues with Realm schema versioning on the app. The backend always had to be compatible with older app versions.

**Reply by hendych** [reply]
Nice insights @philipphofmann ! Did you have any issues on DB migration before? Manual checking prone to human errors. I'm thinking a way to minimize it on DB migration Realm which could impact runtime crashes (I don't know if you used SQLite also will crash or not).

I'm interested on how you do integration tests on those DB :)

**Reply by philipphofmann** [reply]
Yes, we sometimes had issues, if I remember correctly, but I can't remember the details as this was already five years ago or so. We did the manual checks on top of the automated ones. 

**For any DB migration testing**

1. You need to structure your code in a way so that you can install any schema version on an actual phone or a simulator. 
2. Add some test data. Maybe you have a data dump as an SQL file or JSON. You could also do this by calling into the database layer with code.
3. Again, structure your code so you can call your migrations individually and then call your desired migration.

Then you can test every migration you want.
